examples/developer-tools/src/index.ts

```diff
@@ -33,9 +38,60 @@ const startBlocks = {
           'INLINE': 'AUTO',
           'CONNECTIONS': 'NONE',
         },
+        'inputs': {
+          'TOOLTIP': {
+            'block': {
+              'type': 'text',
+              'deletable': false,
+              'movable': false,
+              'fields': {
+                'TEXT': '',
+              },
+            },
+          },
+          'HELPURL': {
+            'block': {
+              'type': 'text',
+              'deletable': false,
+              'movable': false,
+              'fields': {
+                'TEXT': '',
+              },
+            },
+          },
+        },
       },
     ],
   },
 };
 
 Blockly.serialization.workspaces.load(startBlocks, mainWorkspace);
+
+let blockName = '';
+
+
+// Whenever the workspace changes meaningfully, update the preview.
+mainWorkspace.addChangeListener((e: Blockly.Events.Abstract) => {
+  // Don't run the code when the workspace finishes loading; we're
+  // already running it once when the application starts.
+  // Don't run the code during drags; we might have invalid state.
+  if (e.isUiEvent || e.type == Blockly.Events.FINISHED_LOADING ||
+      mainWorkspace.isDragging()) {
+    return;
+  }
```

Comment:
Generally more predictable + future proof to use allow lists rather than deny lists for change listeners. E.g. https://github.com/google/blockly/pull/6337

Reply:
I want pretty much all non-UI events though. I feel like this is a standard pattern we use elsewhere, and a list of every non-UI event would be harder to read and scan for accuracy (easy to forget one).